lib/collection/src/hash_ring.rs

```diff
@@ -58,15 +58,27 @@ impl<T: Hash + Copy + PartialEq> HashRing<T> {
     pub fn add(&mut self, shard: T) {
         match self {
             Self::Single(ring) => ring.add(shard),
-            Self::Resharding { old, new } => {
-                if new.get(&shard).is_none() {
-                    old.add(shard);
+            Self::Resharding { new, old: _ } => {
+                if !new.contains_shard(&shard) {
```

Comment:
This check was incorrect.

`get(point_id) -> target_shard` returns the target shard, and we gave it the shard ID as source. I've updated this with a new method actually checking whether the shard is already in the ring.

In practice we should never reach this branch while resharding, so I've added a debug assertion below.